linked_list/main.py

```diff
@@ -43,53 +43,173 @@ class Node(object):
     """
 
-    def __init__(self, elements=None):
-        pass
-
-    def __str__(self):
-        pass
-
-    def __len__(self):
-        pass
-
-    def __iter__(self):
-        pass
-
-    def __getitem__(self, index):
-        pass
-
-    def __add__(self, other):
-        pass
-
-    def __iadd__(self, other):
-        pass
-
-    def __eq__(self, other):
-        pass
-
-    def append(self, elem):
-        pass
-
-    def count(self):
-        pass
-
-    def pop(self, index=None):
-        pass
+    def __init__(self, elements=None): # jon:DONE
+        #This should initialize a node for each item in the list.
+        # then populate 'elem' with the element, and link up the next node.
+
+        self.elements = elements
+        self.start = None
+        self.end = None
+
+        if self.elements:
+            # set up a new list of 1 item using first passed element.
+            self.start = Node(self.elements[0])
+            self.end = self.start
+            # trim off the first item, so we don't repeat it.
+            self.elements = self.elements[1:]
+            # now iterate and populat ethe rest of the list.
+            for item in self.elements:
+                self.append(item)
+
+    def __str__(self): # corey:DONE
+        # per Line 158 of test_main, a list of items with 3 nodes 1, 2, 3
+        #  should return a string representation "[1, 2, 3]"
+        # This should also allow test_main lines 168+ to work.
+        # However, if not, we may need to implement magic function __repr__
+        string_list = []
+        if self.start != None or []:
+            string_list.append(self.start.elem)
+            curr_node = self.start.next
+            while curr_node != self.end:
+                string_list.append(curr_node.elem)
+                curr_node = curr_node.next
+            string_list.append(self.end.elem)
+        return str(string_list)
+            
+    def __len__(self): # jon:DONE
+        # should simply return the number of nodes.
+        # iterate over list, and return the number of nodes.
+        # we can call count() to count up the nodes.
+        return self.count()
+
+    def __iter__(self): # jon:DONE
+        # start at head.  yield item, increment to next node in list.
+        current = self.start
+        while current:
+            yield current
+            current = current.next
+
+    def __getitem__(self, index): # jon:DONE
+        # fetch item at position index
+        # this is what returns self[key]
+        # TODO: BOUNDS CHECKING!  RAISE OUT OF INDEX ERRORS!
+        current = self.start
+        count = 0
+        for count in range(index):
+            current = current.next
+
+        return current
+
+    def __add__(self, other): # akshith: Done
+        res = []
+        for item in self:
+            res.append(item.elem)
+        for item in other:
+            res.append(item.elem)
+        return LinkedList(res)
```

Comment:
I think this method is relying too much in the use of python lists. I'd try to do it without `res = []`